cloudformation/cloudformation.yaml

```diff
@@ -8,7 +8,109 @@ Parameters:
             - PROD
             - CODE
         Default: CODE
+Mappings:
+  EnvironmentBucketMap:
+    CODE:
+      "logs": "fulfilment-s3-logs-code"
+      "salesforce": "fulfilment-salesforce-backup-code"
+      "zuora": "fulfilment-zuora-staging-code"
+      "export": "fulfilment-export-code"
+    PROD:
+      "logs": "fulfilment-s3-logs-prod"
+      "salesforce": "fulfilment-salesforce-backup-prod"
+      "zuora": "fulfilment-zuora-staging-prod"
+      "export": "fulfilment-export-prod"
 Resources:
+  FulfilmentAccessLogBucket:
+    Type: "AWS::S3::Bucket"
+    Properties:
+      BucketName:
+        Fn::FindInMap:
+          - EnvironmentBucketMap
+          - Ref: Stage
+          - "logs"
+      AccessControl: LogDeliveryWrite
+      LoggingConfiguration:
+        DestinationBucketName:
+          Fn::FindInMap:
+            - EnvironmentBucketMap
+            - Ref: Stage
+            - "logs"
+        LogFilePrefix: "logs-logs/"
+      VersioningConfiguration:
+        Status: "Enabled"
+  SalesForceBackupBucket:
+    Type: "AWS::S3::Bucket"
+    Properties:
+      BucketName:
+        Fn::FindInMap:
+          - EnvironmentBucketMap
+          - Ref: Stage
+          - "salesforce"
+      AccessControl: Private
+      LifecycleConfiguration:
+        Rules:
+          - Id: DeleteOldFilesSF
+            Prefix: ""
+            Status: Enabled
+            ExpirationInDays: 21
+      LoggingConfiguration:
+        DestinationBucketName:
+          Fn::FindInMap:
+            - EnvironmentBucketMap
+            - Ref: Stage
+            - "logs"
+        LogFilePrefix: "fulfilment-salesforce-backup-_[STAGE]/"
+      VersioningConfiguration:
+        Status: "Enabled"
+  ZuoraExportBucket:
+    Type: "AWS::S3::Bucket"
+    Properties:
+      BucketName:
+        Fn::FindInMap:
+          - EnvironmentBucketMap
+          - Ref: Stage
+          - "zuora"
+      AccessControl: Private
+      LifecycleConfiguration:
+        Rules:
+          - Id: DeleteOldFiles
+            Prefix: ""
+            Status: Enabled
+            ExpirationInDays: 21
+      LoggingConfiguration:
+        DestinationBucketName:
+          Fn::FindInMap:
+            - EnvironmentBucketMap
+            - Ref: Stage
+            - "logs"
+        LogFilePrefix: "fulfilment-salesforce-backup-_[STAGE]/"
+      VersioningConfiguration:
+        Status: "Enabled"
+  FulfilmentBucket:
+    Type: "AWS::S3::Bucket"
+    Properties:
+      BucketName:
+        Fn::FindInMap:
+          - EnvironmentBucketMap
+          - Ref: Stage
+          - "export"
+      AccessControl: Private
+      LifecycleConfiguration:
+        Rules:
```

Comment:
Minor point, but we effectively use the same rule 3 times - is it possible to define it once and re-use it?

Reply:
Ugh, I'm sorry. There's a lot of stuff that stops you reusing stuff between buckets, so I refactored to use one and add an encryption policy. (And forgot to push)

Reply:
maybe we could define a [deletion policy](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-deletionpolicy.html) to ensure the buckets are not deleted if the stack is removed or modified? 
Although according to the [doc](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket.html) it looks like the default behaviour is to fail the stack update if a non empty bucket needs to be deleted